test/quic_wire_test.c

```diff
@@ -0,0 +1,1224 @@
+/*
+ * Copyright 2015-2018 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include "internal/packet.h"
+#include "internal/quic_wire.h"
+#include "testutil.h"
+
+struct encode_test_case {
+    int (*serializer)(WPACKET *pkt);
+    const uint8_t *expect_buf;
+    size_t expect_buf_len;
+    int (*deserializer)(PACKET *pkt);
+};
+
+/* 1. PADDING */
+static int encode_case_1_enc(WPACKET *pkt)
+{
+    if (!TEST_int_eq(ossl_quic_wire_encode_frame_padding(pkt, 3), 1))
+        return 0;
+
+    return 1;
+}
+
+static int encode_case_1_dec(PACKET *pkt)
+{
+    if (!TEST_int_eq(ossl_quic_wire_decode_frame_padding(pkt), 3))
+        return 0;
+
+    return 1;
+}
+
+static const uint8_t encode_case_1_expect[] = {
+    0, 0, 0
+};
+
+/* 2. PING */
+static int encode_case_2_enc(WPACKET *pkt)
+{
+
+    if (!TEST_int_eq(ossl_quic_wire_encode_frame_ping(pkt), 1))
+        return 0;
+
+    return 1;
+}
+
+static int encode_case_2_dec(PACKET *pkt)
+{
+
+    if (!TEST_int_eq(ossl_quic_wire_decode_frame_ping(pkt), 1))
+        return 0;
+
+    return 1;
+}
+
+static const uint8_t encode_case_2_expect[] = {
+    0x01
+};
+
+/* 3. ACK */
```

Comment:
I feel like there are perhaps more test cases required here:

E.g. ack frame with no gaps, maximal/minimal ack delay values, no ecn data, etc

Reply:
The ACK frames are definitely the hairiest frames to encode and decode. There are more tests specifically for ACK below, but more would probably be good.

ACKs with and without ECN are tested, but we can probably test ACK delays more. I'll take this as an action item.